modules/tfm/tfm/boards/src/log_memory_protection.c

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <cmsis.h>
+#include <tfm_spm_log.h>
+#include <mpu_armv8m_drv.h>
+#include <nrf.h>
+#include <nrf_peripherals.h>
+#include <array.h>
+
+
+void log_memory_protection_of_mpu(void)
+{
+	MPU_Type * mpu = (MPU_Type *)MPU_BASE;
+	SPMLOG_DBGMSG("ARM MPU configuration\r\n");
+	SPMLOG_DBGMSGVAL("CTRL", mpu->CTRL);
+
+	for(int i = 0; i < MPU_ARMV8M_NUM_REGIONS; i++) {
+		mpu->RNR = i;
+		SPMLOG_DBGMSGVAL("RNR", i);
+		uint32_t rbar = mpu->RBAR;
+		uint32_t rlar = mpu->RLAR;
+		uint32_t en = (rlar & MPU_RLAR_EN_Msk) >> MPU_RLAR_EN_Pos;
+
+		SPMLOG_DBGMSGVAL("  EN", en);
+		if(en) {
+			SPMLOG_DBGMSGVAL("  BASE", rbar & MPU_RBAR_BASE_Msk);
+
+			// The last address that is protected is limit + 31
+			SPMLOG_DBGMSGVAL("  LIMIT*",  (rlar & MPU_RLAR_LIMIT_Msk) + 31);
+
+			SPMLOG_DBGMSGVAL("  XN",  (rbar & MPU_RBAR_XN_Msk) >> MPU_RBAR_XN_Pos);
+			SPMLOG_DBGMSGVAL("  AP",  (rbar & MPU_RBAR_AP_Msk) >> MPU_RBAR_AP_Pos);
+			SPMLOG_DBGMSGVAL("  SH",  (rbar & MPU_RBAR_SH_Msk) >> MPU_RBAR_SH_Pos);
+
+			SPMLOG_DBGMSGVAL("  ATTRINDEX",  (rlar & MPU_RLAR_AttrIndx_Msk) >> MPU_RLAR_AttrIndx_Pos);
+		}
+	}
+}
+
+void log_memory_protection_of_spu(void)
+{
+	SPMLOG_DBGMSGVAL("NRF_SPU_S->FLASHNSC[0].REGION", NRF_SPU_S->FLASHNSC[0].REGION);
+	SPMLOG_DBGMSGVAL("NRF_SPU_S->FLASHNSC[1].REGION", NRF_SPU_S->FLASHNSC[1].REGION);
+
+	SPMLOG_DBGMSGVAL("NRF_SPU_S->FLASHNSC[0].SIZE", NRF_SPU_S->FLASHNSC[0].SIZE);
+	SPMLOG_DBGMSGVAL("NRF_SPU_S->FLASHNSC[1].SIZE", NRF_SPU_S->FLASHNSC[1].SIZE);
+
+	SPMLOG_DBGMSGVAL("NRF_SPU_S->RAMNSC[0].REGION", NRF_SPU_S->RAMNSC[0].REGION);
+	SPMLOG_DBGMSGVAL("NRF_SPU_S->RAMNSC[1].REGION", NRF_SPU_S->RAMNSC[1].REGION);
+
+	SPMLOG_DBGMSGVAL("NRF_SPU_S->RAMNSC[0].SIZE", NRF_SPU_S->RAMNSC[0].SIZE);
+	SPMLOG_DBGMSGVAL("NRF_SPU_S->RAMNSC[1].SIZE", NRF_SPU_S->RAMNSC[1].SIZE);
+
+
+#if defined(NRF9160_XXAA) || defined(NRF5340_XXAA_APPLICATION)
+	// Scary, maybe there will be a variant with a differently sized
+	// flash in the future.
+#define NVMC_FLASH_SIZE (1024 * 1024) // 1 MB
```

Comment:
This should be read using FICR->INFO, Total code space is: CODEPAGESIZE * CODESIZE bytes

Reply:
done